source/fuzz/transformation_propagate_instruction_up.cpp

```diff
@@ -0,0 +1,397 @@
+                     });
+}
+
+void TransformationPropagateInstructionUp::Apply(
+    opt::IRContext* ir_context, TransformationContext* /*unused*/) const {
+  auto* inst = GetInstructionToPropagate(ir_context, message_.block_id());
+  assert(inst &&
+         "The block must have at least one supported instruction to propagate");
+
+  // |inst| might depend on OpPhi instructions from the same basic block.
+  // That's OK, because by looking at the OpPhi instruction we can tell which of
+  // its operands we should use instead when we propagate the instruction into
+  // each predecessor.
+  //
+  // |op_phi_to_result_id| contains a mapping from the result id of such an
+  // OpPhi instruction to the map of its operands
+  // (i.e. |op_phi_to_result_id[op_phi_id][label_id] == result_id|).
+  const auto op_phi_to_result_id =
+      ComputeMappingFromOpPhiToResultId(ir_context, inst);
+
+  opt::Instruction::OperandList op_phi_operands;
+  const auto predecessor_id_to_fresh_id = fuzzerutil::RepeatedUInt32PairToMap(
+      message_.predecessor_id_to_fresh_id());
+  for (auto predecessor_id : ir_context->cfg()->preds(message_.block_id())) {
+    auto new_result_id = predecessor_id_to_fresh_id.at(predecessor_id);
+
+    // Compute InOperands for the OpPhi instruction to be inserted later.
+    op_phi_operands.push_back({SPV_OPERAND_TYPE_ID, {new_result_id}});
+    op_phi_operands.push_back({SPV_OPERAND_TYPE_ID, {predecessor_id}});
+
+    // Create a clone of the |inst| to be inserted into the |predecessor_id|.
+    std::unique_ptr<opt::Instruction> clone(inst->Clone(ir_context));
+    clone->SetResultId(new_result_id);
+
+    fuzzerutil::UpdateModuleIdBound(ir_context, new_result_id);
+
+    // Adjust |clone|'s operands to account for possible dependencies on OpPhi
+    // instructions from the same basic block.
+    for (uint32_t i = 0; i < clone->NumInOperands(); ++i) {
+      auto& operand = clone->GetInOperand(i);
+
+      if (operand.type == SPV_OPERAND_TYPE_ID &&
+          op_phi_to_result_id.count(operand.words[0])) {
+        operand.words[0] =
+            op_phi_to_result_id.at(operand.words[0]).at(predecessor_id);
+      }
+    }
+
+    // Insert cloned instruction into the predecessor.
+    auto* predecessor = ir_context->cfg()->block(predecessor_id);
+    assert(predecessor && "|predecessor_id| is invalid");
+
+    GetLastInsertBeforeInstruction(predecessor, clone->opcode())
+        ->InsertBefore(std::move(clone));
+  }
+
+  // Insert an OpPhi instruction into the basic block of |inst|.
+  ir_context->get_instr_block(inst)->begin()->InsertBefore(
+      MakeUnique<opt::Instruction>(ir_context, SpvOpPhi, inst->type_id(),
+                                   inst->result_id(),
+                                   std::move(op_phi_operands)));
+
+  // Remove |inst| from the basic block.
+  inst->RemoveFromList();
+  delete inst;  // RemoveFromList doesn't clear the memory.
```

Comment:
If it is not too painful, maybe keep an iterator to the instruction (instead of a pointer to the instruction) and then instead use `->Erase()`.

Reply:
Actually, ignore this, as it seems like you have to use delete below anyway.

Reply:
Using `KillInst` in dfd7889.